fairscale/optim/oss.py

```diff
@@ -201,6 +202,72 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
 
         return loss
 
+    def clip_grad_norm(self, max_norm: float, norm_type: float = 2.0) -> torch.Tensor:
+        """
+        Clip all gradients at this point in time. The norm is computed over all gradients together, as if they were
+        concatenated into a single vector. Gradients are modified in-place.
+
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for
+                infinity norm.
+
+        Returns:
+            Total norm of the parameters (viewed as a single vector).
+
+        .. note: This is analogous to `torch.nn.utils.clip_grad_norm_` but handles the partitioning and multiple devices per rank
+        under the hood.
+
+        .. warning: This needs to be called on all ranks, since synchronization primitives will be used
+        """
+
+        # Compute the max norm for this shards's worth of gradients
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+
+        # Filter out the grad-less params, concatenate params from all devices
+        local_params = [
+            filter(lambda x: x.grad is not None, device_params[self.rank])
+            for device_params in self.per_device_params.values()
+        ]
+
+        # Compute the norm on this grad set
+        if norm_type == inf:
+            local_norm = max(
+                p.grad.detach().abs().max().to(self._device)  # type: ignore   # mypy seems wrong here, None grads have been filtered out
```

Comment:
Hey @blefaudeux, this implementation probably won't work for model parallelism - the gradients need to be all reduced before applying this operation.

Are you planning to maintain support for model parallelism? I'm happy to contribute a PR! One approach is to mark each parameter in the `mpu` with a bit showing whether it's part of model parallelism and then allreduce just those gradients.

Reply:
I've little visibility in the model parallelism part, I'm really trying to keep it into account but as you can see it's not obvious. I did have a specific DDP in mind when writing that (for instance #157), which removes gradients from shards which don't own it and auto-reduce (not all-reduce), I realize now that the computation is wrong if it's not the case but then the interface will get a little clumsy